src/browser/extension/options/syncOptions.js

```diff
@@ -1,39 +1,64 @@
+import { FilterState } from '../utils/filters';
+
 let options;
+let subscribers = [];
 
 const save = (toAllTabs) => (key, value) => {
   let obj = {};
   obj[key] = value;
   chrome.storage.sync.set(obj);
   options[key] = value;
   toAllTabs({ options: options });
+  subscribers.forEach(s => s(options));
+};
+
+const migrateOldOptions = (oldOptions) => {
+  let newOptions = Object.assign({}, oldOptions);
+
+  // Migrate the old `filter` option from 2.2.1
+  if (typeof oldOptions.filter === 'boolean') {
+    if (oldOptions.filter && oldOptions.whitelist.length > 0) {
+      newOptions.filter = FilterState.WHITELIST_SPECIFIC;
+    } else if (oldOptions.filter) {
+      newOptions.filter = FilterState.BLACKLIST_SPECIFIC;
+    } else {
+      newOptions.filter = FilterState.DO_NOT_FILTER;
+    }
+  }
+
+  return newOptions;
 };
 
 const get = callback => {
   if (options) callback(options);
   else {
     chrome.storage.sync.get({
       maxAge: 50,
-      filter: false,
+      filter: FilterState.DO_NOT_FILTER,
       whitelist: '',
       blacklist: '',
       serialize: true,
       notifyErrors: true,
       inject: true,
       urls: '^https?://localhost|0\\.0\\.0\\.0:\\d+\n^https?://.+\\.github\\.io'
     }, function(items) {
-      options = items;
-      callback(items);
+      options = migrateOldOptions(items);
```

Comment:
In the new design, the `filter` option isn’t Boolean anymore and can have three values: “Don’t filter”, “Filter whitelisted” and “Filter blacklisted”. The old `options` structure was inappropriate for this, so I added a method to migrate the old options to the new format. This makes the update painless for the users.